trinity/protocol/les/managers.py

```diff
@@ -41,36 +46,58 @@
 ]
 
 
-class GetBlockHeadersRequestManager(BaseRequestManager):
+class HeaderRequestArgs(NamedTuple):
+    block_number_or_hash: BlockIdentifier
+    max_headers: int
+    skip: int
+    reverse: bool
+    request_id: int
+
+
+BlockHeadersRequestManager = RequestManager['LESPeer', Dict[str, Any]]
+
+
+class GetBlockHeadersRequestManager(BaseService):
     msg_queue_maxsize = 100
 
-    _response_msg_type: Type[Command] = BlockHeaders
+    def __init__(self, peer: 'LESPeer', token: CancelToken) -> None:
+        super().__init__(token)
+        self._peer = peer
+        self._request_manager = BlockHeadersRequestManager(peer, BlockHeaders, token)
 
-    # All `RequestManager` classes are expected to implement the `__call__`
-    # method, including changing the function signature, thus the
-    # `# type: ignore` here is both expected and required.
-    async def __call__(self,  # type: ignore
+    async def __call__(self,
                        block_number_or_hash: BlockIdentifier,
                        max_headers: int = None,
                        skip: int = 0,
                        reverse: bool = True,
                        timeout: int = None) -> Tuple[BlockHeader, ...]:
-        request_id = gen_request_id()
-        request = HeaderRequest(
+        request_args = HeaderRequestArgs(
             block_number_or_hash,
             max_headers,
             skip,
             reverse,
-            request_id,
+            gen_request_id(),
         )
-        return await self._request_and_wait(request, timeout)
+        make_request = partial(self._peer.sub_proto.send_get_block_headers, *request_args)
+        validator = HeaderRequest(*request_args)
+        manager = self._request_manager
+
+        async for message in manager.message_candidates(make_request, self.logger, timeout):
+            async with disconnect_on_misbehavior(self._peer, BlockHeaders, self.logger):
+                result = self._normalize_result(message)
+
+            item_count = len(message['headers'])
+
+            with manager.complete_if_valid(item_count):
```

Comment:
I am currently :-1: on this API.  It feels very *magic*.  Basic `try/except` handling here seems like it would be significantly easier to read/understand since it should be more explicit, but it shouldn't be that much extra code.